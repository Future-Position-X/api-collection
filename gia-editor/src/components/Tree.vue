<template>
  <CollectionTreeView
      ref="treeview"
      return-object
      @input="onSelectionUpdate"
      @update:active="onActiveUpdate"
      :items="items"
      selectable
      activatable
      :open.sync="open"
  >
    <template v-slot:append="{ item }">
      <v-icon :color="item.color">mdi-brightness-1</v-icon>
    </template>
  </CollectionTreeView>
</template>

<script>
import CollectionTreeView from "./CollectionTreeView.vue";
import session from '../services/session';
import providerService from '../services/provider';

export default {
  components: {CollectionTreeView},
  props: ['sortedCollections'],
  methods: {
    click() {
    },
    onActiveUpdate(activeCollections) {
      console.debug("onActiveUpdate", activeCollections);
      if (activeCollections.length >= 1) {
        const collection = activeCollections[0];
        this.$emit('activeUpdate', collection.uuid)
      } else {
        this.$emit('activeUpdate', null)
      }
    },
    onSelectionUpdate(selectedItems) {
      console.debug("onSelectionUpdate", selectedItems);
      const selectedUuids = selectedItems.filter((x) => !!x.uuid).map((x) => x.uuid);
      console.debug("selectedUuids: ", selectedUuids);
      this.$emit('selectionUpdate', selectedUuids);
    },
    addCollection(collection) {
      const newCollection = {...collection, editable: true, activatable: true, selectable: true, id: collection.uuid};

      console.debug("newCollection: ", newCollection);
      const item = {
        id: newCollection.uuid,
        uuid: newCollection.uuid,
        name: newCollection.name,
        color: newCollection.color,
        provider_uuid: newCollection.provider_uuid,
        editable: true,
        activatable: true,
        selectable: true
      };

      this.items[0].children.push(item);
    },
    removeCollection(collection) {
      this.items.splice(this.items.indexOf(collection), 1);
    },
    async cacheProvider(uuid) {
      let provider = this.providerCache[uuid];

      if (provider == null) {
        provider = await providerService.get(uuid);
        this.providerCache[uuid] = provider;
      }
    }
  },
  watch: {
    async sortedCollections() {
      this.items = [];
      const collections = [];

      const promises = Object.values(this.sortedCollections)
        .reduce((acc, coll) => acc.concat(coll))
        .map(c => this.cacheProvider(c.provider_uuid));

      await Promise.all(promises);

      // eslint-disable-next-line no-restricted-syntax
      for (const [key, value] of Object.entries(this.sortedCollections)) {
        collections.push({
          id: key,
          name: key,
          color: '#FFF',
          provider_uuid: value[0].provider_uuid,
          editable: false,
          activatable: false,
          selectable: false,
          children: value.map((e) => {
            e.id = e.uuid;
            e.editable = true;
            e.activatable = true;
            e.selectable = true;
            const suffix = ` (${this.providerCache[e.provider_uuid].name})`
            if (!e.name.endsWith(suffix)) {
              e.name += suffix;
            }
            // e.name = e.provider_uuid;
            return e;
          }),
        });
      }
      let providerUuid = null;
      if (session.authenticated()) {
        providerUuid = session.user.provider_uuid;
      }

      const our = Object.values(this.sortedCollections)
        .reduce((acc, coll) => acc.concat(coll))
        .filter(coll => coll.provider_uuid === providerUuid);

      this.items.push({
        id: 'Owned collections',
        name: 'Owned collections',
        color: '#FFF',
        editable: false,
        activatable: false,
        selectable: false,
        children: our.map(coll => ({
          id: coll.uuid,
          uuid: coll.uuid,
          name: coll.name,
          color: coll.color,
          provider_uuid: providerUuid,
          editable: true,
          activatable: true,
          selectable: true
        }))
      });
      this.items.push({
        id: 'Other collections',
        name: 'Other collections',
        color: '#FFF',
        editable: false,
        activatable: false,
        selectable: false,
        children: collections.filter((coll) => coll.provider_uuid !== providerUuid),
      });
      this.open = ['Owned collections'];
    },
  },
  data: () => ({
    items: [],
    open: [],
    providerCache: {}
  }),
};
</script>
