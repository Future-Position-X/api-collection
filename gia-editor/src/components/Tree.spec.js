import { mount } from '@vue/test-utils';
import Vue from 'vue';
import Vuetify from 'vuetify';

import Tree from './Tree.vue';

Vue.use(Vuetify);

const providerCache = {
  "677c6efd-5a6b-4201-b247-e1068c0376fc": {
    uuid: "677c6efd-5a6b-4201-b247-e1068c0376fc",
    name: "fpx"
  }
};

const sortedCollections = {
  __test: [
    {
      uuid: 'daa0c041-e720-48ee-bd73-657c6ef55d25',
      provider_uuid: '677c6efd-5a6b-4201-b247-e1068c0376fc',
      name: '__test',
      is_public: true,
      revision: 1,
      created_at: '2020-06-23 07:03:06.926628',
      updated_at: '2020-06-23 07:03:06.926628',
      color: 'hsl(7.5,100%,50%)',
      id: 'daa0c041-e720-48ee-bd73-657c6ef55d25',
      editable: true,
    },
  ],
  gg: [
    {
      uuid: '00000000-0000-0000-0000-000000000000',
      provider_uuid: '677c6efd-5a6b-4201-b247-e1068c0376fc',
      name: 'gg',
      is_public: true,
      revision: 0,
      created_at: '2020-05-07 13:07:23.126129',
      updated_at: '2020-05-07 13:07:23.126129',
      color: 'hsl(0,100%,50%)',
      id: '00000000-0000-0000-0000-000000000000',
      editable: true,
    },
  ],
};
const items = [
  {
    id: 'Owned collections',
    name: 'Owned collections',
    color: '#FFF',
    editable: false,
    children: [],
  },
  {
    id: 'Other collections',
    name: 'Other collections',
    color: '#FFF',
    editable: false,
    children: [
      {
        id: '__test',
        name: '__test',
        color: '#FFF',
        provider_uuid: '677c6efd-5a6b-4201-b247-e1068c0376fc',
        editable: false,
        children: [
          {
            uuid: 'daa0c041-e720-48ee-bd73-657c6ef55d25',
            provider_uuid: '677c6efd-5a6b-4201-b247-e1068c0376fc',
            name: '__test (fpx)',
            is_public: true,
            revision: 1,
            created_at: '2020-06-23 07:03:06.926628',
            updated_at: '2020-06-23 07:03:06.926628',
            color: 'hsl(7.5,100%,50%)',
            id: 'daa0c041-e720-48ee-bd73-657c6ef55d25',
            editable: true,
          },
        ],
      },
      {
        id: 'gg',
        name: 'gg',
        color: '#FFF',
        provider_uuid: '677c6efd-5a6b-4201-b247-e1068c0376fc',
        editable: false,
        children: [
          {
            uuid: '00000000-0000-0000-0000-000000000000',
            provider_uuid: '677c6efd-5a6b-4201-b247-e1068c0376fc',
            name: 'gg (fpx)',
            is_public: true,
            revision: 0,
            created_at: '2020-05-07 13:07:23.126129',
            updated_at: '2020-05-07 13:07:23.126129',
            color: 'hsl(0,100%,50%)',
            id: '00000000-0000-0000-0000-000000000000',
            editable: true,
          },
        ],
      },
    ],
  },
];

describe('Tree', () => {
  describe('data', () => {
    it('should render no elements when tree is empty', async () => {
      const wrapper = mount(Tree);
      expect(wrapper.element).toMatchSnapshot();
    });

    it('should render "owned" and "other" root elems when tree has items', async () => {
      const wrapper = mount(Tree);
      await wrapper.setData({ providerCache });
      await wrapper.setProps({ sortedCollections });
      expect(wrapper.vm.items).toMatchObject(items);
      expect(wrapper.element).toMatchSnapshot();
    });

    it('should render child items when opened', async () => {
      const wrapper = mount(Tree);
      await wrapper.setData({ providerCache });
      await wrapper.setProps({ sortedCollections });
      await wrapper.findComponent({ name: 'v-treeview' }).vm.updateOpen('Owned collections', false);
      await wrapper.findComponent({ name: 'v-treeview' }).vm.updateOpen('Other collections', true);
      expect(wrapper.text()).toContain('__test');
      expect(wrapper.text()).toContain('gg');
      expect(wrapper.element).toMatchSnapshot();
    });

    it('should render $checkboxOn when item is selected', async () => {
      const wrapper = mount(Tree);
      await wrapper.setData({ providerCache });
      await wrapper.setProps({ sortedCollections });
      const treeview = await wrapper.findComponent({ name: 'v-treeview' }).vm;
      await treeview.updateOpen('Owned collections', false);
      await treeview.updateOpen('Other collections', true);
      await treeview.updateOpen('__test', true);
      await treeview.updateSelected('daa0c041-e720-48ee-bd73-657c6ef55d25', true);
      expect(wrapper.text()).toContain('$checkboxOn__test (fpx)$subgroupgg');
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('selectionUpdate event', () => {
    it('should not include collections without uuids', async () => {
      // If "Owned collections" is empty and selected then it is emitted from v-treeview
      const wrapper = mount(Tree);
      await wrapper.setProps({ sortedCollections });
      await wrapper.findComponent({ name: 'v-treeview' }).vm.$emit('input', [items[0]]);
      const selectionUpdateEvents = wrapper.emitted('selectionUpdate');
      const firstEvent = selectionUpdateEvents[0];
      const firstArgument = firstEvent[0];
      expect(firstArgument).toEqual([]);
    });

    it('should include collections with uuids', async () => {
      const wrapper = mount(Tree);
      await wrapper.setProps({ sortedCollections });
      await wrapper
        .findComponent({ name: 'v-treeview' })
        .vm.$emit('input', Object.values(sortedCollections).flat());
      const selectionUpdateEvents = wrapper.emitted('selectionUpdate');
      const firstEvent = selectionUpdateEvents[0];
      const firstArgument = firstEvent[0];
      expect(firstArgument).toEqual([
        'daa0c041-e720-48ee-bd73-657c6ef55d25',
        '00000000-0000-0000-0000-000000000000',
      ]);
    });
  });
});
