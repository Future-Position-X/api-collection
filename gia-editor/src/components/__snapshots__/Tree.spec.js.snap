// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree data should render "owned" and "other" root elems when tree has items 1`] = `
<div
  class="v-treeview theme--light"
>
  <div
    aria-expanded="false"
    class="v-treeview-node v-treeview-node--leaf"
  >
    <div
      class="v-treeview-node__root"
    >
      <div
        class="v-treeview-node__level"
      />
      <button
        class="v-icon notranslate v-treeview-node__checkbox v-icon--link material-icons theme--light"
        type="button"
      >
        $checkboxOff
      </button>
      <div
        class="v-treeview-node__content"
      >
        <div
          class="v-treeview-node__label"
        >
          Owned collections
        </div>
        <div
          class="v-treeview-node__append"
        >
          <i
            aria-hidden="true"
            class="v-icon notranslate mdi mdi-brightness-1 theme--light"
            style="color: rgb(255, 255, 255); caret-color: #FFF;"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    aria-expanded="false"
    class="v-treeview-node"
  >
    <div
      class="v-treeview-node__root"
    >
      <button
        class="v-icon notranslate v-treeview-node__toggle v-icon--link material-icons theme--light"
        type="button"
      >
        $subgroup
      </button>
      <button
        class="v-icon notranslate v-treeview-node__checkbox v-icon--link material-icons theme--light"
        type="button"
      >
        $checkboxOff
      </button>
      <div
        class="v-treeview-node__content"
      >
        <div
          class="v-treeview-node__label"
        >
          Other collections
        </div>
        <div
          class="v-treeview-node__append"
        >
          <i
            aria-hidden="true"
            class="v-icon notranslate mdi mdi-brightness-1 theme--light"
            style="color: rgb(255, 255, 255); caret-color: #FFF;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Tree data should render $checkboxOn when item is selected 1`] = `
<div
  class="v-treeview theme--light"
>
  <div
    aria-expanded="false"
    class="v-treeview-node v-treeview-node--leaf"
  >
    <div
      class="v-treeview-node__root"
    >
      <div
        class="v-treeview-node__level"
      />
      <button
        class="v-icon notranslate v-treeview-node__checkbox v-icon--link material-icons theme--light"
        type="button"
      >
        $checkboxOff
      </button>
      <div
        class="v-treeview-node__content"
      >
        <div
          class="v-treeview-node__label"
        >
          Owned collections
        </div>
        <div
          class="v-treeview-node__append"
        >
          <i
            aria-hidden="true"
            class="v-icon notranslate mdi mdi-brightness-1 theme--light"
            style="color: rgb(255, 255, 255); caret-color: #FFF;"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    aria-expanded="false"
    class="v-treeview-node v-treeview-node--selected"
  >
    <div
      class="v-treeview-node__root"
    >
      <button
        class="v-icon notranslate v-treeview-node__toggle v-icon--link material-icons theme--light"
        type="button"
      >
        $subgroup
      </button>
      <button
        class="v-icon notranslate v-treeview-node__checkbox v-icon--link material-icons theme--light accent--text"
        type="button"
      >
        $checkboxOn
      </button>
      <div
        class="v-treeview-node__content"
      >
        <div
          class="v-treeview-node__label"
        >
          Other collections
        </div>
        <div
          class="v-treeview-node__append"
        >
          <i
            aria-hidden="true"
            class="v-icon notranslate mdi mdi-brightness-1 theme--light"
            style="color: rgb(255, 255, 255); caret-color: #FFF;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Tree data should render child items when opened 1`] = `
<div
  class="v-treeview theme--light"
>
  <div
    aria-expanded="false"
    class="v-treeview-node v-treeview-node--leaf"
  >
    <div
      class="v-treeview-node__root"
    >
      <div
        class="v-treeview-node__level"
      />
      <button
        class="v-icon notranslate v-treeview-node__checkbox v-icon--link material-icons theme--light"
        type="button"
      >
        $checkboxOff
      </button>
      <div
        class="v-treeview-node__content"
      >
        <div
          class="v-treeview-node__label"
        >
          Owned collections
        </div>
        <div
          class="v-treeview-node__append"
        >
          <i
            aria-hidden="true"
            class="v-icon notranslate mdi mdi-brightness-1 theme--light"
            style="color: rgb(255, 255, 255); caret-color: #FFF;"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    aria-expanded="true"
    class="v-treeview-node"
  >
    <div
      class="v-treeview-node__root"
    >
      <button
        class="v-icon notranslate v-treeview-node__toggle v-icon--link material-icons theme--light v-treeview-node__toggle--open"
        type="button"
      >
        $subgroup
      </button>
      <button
        class="v-icon notranslate v-treeview-node__checkbox v-icon--link material-icons theme--light"
        type="button"
      >
        $checkboxOff
      </button>
      <div
        class="v-treeview-node__content"
      >
        <div
          class="v-treeview-node__label"
        >
          Other collections
        </div>
        <div
          class="v-treeview-node__append"
        >
          <i
            aria-hidden="true"
            class="v-icon notranslate mdi mdi-brightness-1 theme--light"
            style="color: rgb(255, 255, 255); caret-color: #FFF;"
          />
        </div>
      </div>
    </div>
    <div
      class="v-treeview-node__children"
    >
      <div
        aria-expanded="false"
        class="v-treeview-node"
      >
        <div
          class="v-treeview-node__root"
        >
          <div
            class="v-treeview-node__level"
          />
          <button
            class="v-icon notranslate v-treeview-node__toggle v-icon--link material-icons theme--light"
            type="button"
          >
            $subgroup
          </button>
          <button
            class="v-icon notranslate v-treeview-node__checkbox v-icon--link material-icons theme--light"
            type="button"
          >
            $checkboxOff
          </button>
          <div
            class="v-treeview-node__content"
          >
            <div
              class="v-treeview-node__label"
            >
              __test
            </div>
            <div
              class="v-treeview-node__append"
            >
              <i
                aria-hidden="true"
                class="v-icon notranslate mdi mdi-brightness-1 theme--light"
                style="color: rgb(255, 255, 255); caret-color: #FFF;"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        aria-expanded="false"
        class="v-treeview-node"
      >
        <div
          class="v-treeview-node__root"
        >
          <div
            class="v-treeview-node__level"
          />
          <button
            class="v-icon notranslate v-treeview-node__toggle v-icon--link material-icons theme--light"
            type="button"
          >
            $subgroup
          </button>
          <button
            class="v-icon notranslate v-treeview-node__checkbox v-icon--link material-icons theme--light"
            type="button"
          >
            $checkboxOff
          </button>
          <div
            class="v-treeview-node__content"
          >
            <div
              class="v-treeview-node__label"
            >
              gg
            </div>
            <div
              class="v-treeview-node__append"
            >
              <i
                aria-hidden="true"
                class="v-icon notranslate mdi mdi-brightness-1 theme--light"
                style="color: rgb(255, 255, 255); caret-color: #FFF;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Tree data should render no elements when tree is empty 1`] = `
<div
  class="v-treeview theme--light"
/>
`;
